services/venice-controller/src/main/java/com/linkedin/venice/controller/StoreGraveyardCleanupService.java

```diff
@@ -0,0 +1,109 @@
+ * Each child controller checks if the store can be removed from graveyard in the child colo. If so, remove it and
+ * respond OK. Otherwise, respond with error. If all child controllers respond OK, parent controller will remove
+ * the store from graveyard in parent colo.
+ *
+ * The service only removes batch-only stores for now because hybrid ETL needs graveyard to guarantee version numbers
+ * are not reused in recreated stores before ETL off-boarding steps.
+ */
+public class StoreGraveyardCleanupService extends AbstractVeniceService {
+  private static final Logger LOGGER = LogManager.getLogger(StoreGraveyardCleanupService.class);
+
+  private final VeniceParentHelixAdmin admin;
+  private final VeniceControllerMultiClusterConfig multiClusterConfig;
+  private final Thread cleanupThread;
+  private final long sleepIntervalBetweenListFetchMs; // default is 15 min
+  private final long sleepIntervalBetweenStore = TimeUnit.MINUTES.toMillis(1);
+  private final Time time = new SystemTime();
+  private boolean stop = false;
+
+  public StoreGraveyardCleanupService(
+      VeniceParentHelixAdmin admin,
+      VeniceControllerMultiClusterConfig multiClusterConfig) {
+    this.admin = admin;
+    this.multiClusterConfig = multiClusterConfig;
+    this.sleepIntervalBetweenListFetchMs = multiClusterConfig.getGraveyardCleanupSleepIntervalBetweenListFetchMs();
+    this.cleanupThread = new Thread(new StoreGraveyardCleanupTask(), "StoreGraveyardCleanupTask");
+  }
+
+  @Override
+  public boolean startInner() throws Exception {
+    cleanupThread.start();
+    return true;
+  }
+
+  @Override
+  public void stopInner() throws Exception {
+    stop = true;
+    cleanupThread.interrupt();
+  }
+
+  private class StoreGraveyardCleanupTask implements Runnable {
+    @Override
+    public void run() {
+      LOGGER.info("Started running {}", getClass().getSimpleName());
+      while (!stop) {
+        try {
+          time.sleep(sleepIntervalBetweenListFetchMs);
+          // loop all clusters
+          for (String clusterName: multiClusterConfig.getClusters()) {
+            boolean cleanupForBatchOnlyStoreEnabled =
+                multiClusterConfig.getControllerConfig(clusterName).isGraveyardCleanupForBatchOnlyStoreEnabled();
+            if (!cleanupForBatchOnlyStoreEnabled || !admin.isLeaderControllerFor(clusterName)) {
+              // Only clean up when config is enabled and this is the leader controller for current cluster
+              continue;
+            }
+            // List all stores from graveyard for current cluster
+            List<String> storeNames = admin.getStoreGraveyard().listStoreNamesFromGraveyard(clusterName);
+            for (String storeName: storeNames) {
+              boolean didCleanup = false;
+              try {
+                Store store = admin.getStoreGraveyard().getStoreFromGraveyard(clusterName, storeName);
+                if (store != null && !store.isHybrid() && !store.isIncrementalPushEnabled()) {
```

Comment:
Today inc push stores are all hybrid stores. But in store graveyard, there are old inc push stores that are not hybrid (because previously inc push store could be pushed to VT and did not to be hybrid). Since this commit is limited to delete batch-only store graveyard, to align with the commit message, I also skip inc push + non hybrid store.

Reply:
Is there any concern with deleting old non-hybrid inc push stores? If not, we can accommodate them in this commit as they are functionally equivalent to batch only stores from store graveyard's perspective

Reply:
I added a configurable delay so that hybrid store graveyard cleanup is now supported too.